luaui/Widgets/gui_top_bar.lua

```diff
@@ -1087,29 +1849,351 @@ function widget:GameStart()
+					usedBPMetalExpense = usedBPMetalExpense + currentlyUsedM
+					usedBPEnergyExpense = usedBPEnergyExpense + currentlyUsedE -- A builder may be cloaked, but not while it's building
+
+					--currentlyUsedBP = (Spring.GetUnitCurrentBuildPower(unitID) or 0) * currentUnitBP
+					currentlyUsedBP = currentlyUsedM / unitCostData[builtUnitDefID].MperBP -- everything costs at least 1 metal
+					buildingBP = buildingBP + currentUnitBP
+
+					local currentlyWantedM = unitCostData[builtUnitDefID].MperBP * currentUnitBP
+					local currentlyWantedE = unitCostData[builtUnitDefID].EperBP * currentUnitBP
+
+					-- Low-priority units don't have their pulled M/E reported correctly.
+
+					local nonStalledRateM = 1
+					local nonStalledRateE = 1
+					if currentlyWantedM > 0 then
+						nonStalledRateM = currentlyUsedM / currentlyWantedM
+					end
+					if currentlyWantedE > 0 then
+						nonStalledRateE = currentlyUsedE / currentlyWantedE
+					end
+					nonStalledBuildingBP = nonStalledBuildingBP + currentUnitBP * math_min(nonStalledRateM, nonStalledRateE)  -- ROBERT BP at the right place
+
+					if currentlyUsedBP and currentlyUsedBP >= 0 then
+						cacheTotallyUsedBP = cacheTotallyUsedBP + currentlyUsedBP
+					end
+				end
+			end
+			nowChecking = nowChecking + 1
+			--LogFrame("Incremented nowChecking: " .. nowChecking)
+			----LogFrame("nowChecking new one" ..nowChecking)
+		end
+		--LogFrame("Finished trackedBuilders loop")
+		for unitID, unitReservedBP in pairs(unitsReservedBP) do
+			cacheTotalReservedBP = cacheTotalReservedBP + unitReservedBP
+		end
+		cacheDataBase[3] = cacheDataBase[3] + cacheTotalReservedBP
+		cacheDataBase[5] = cacheDataBase[5] + cacheTotallyUsedBP
+		-- TODO: consider averaging income and expense across all frames for which we calculated builder data.
+		local metal, metalStorage, metalPull, metalIncome, metalExpense, metalShare, metalSent = spGetTeamResources(myTeamID, "metal")
+		local energy, energyStorage, energyPull, energyIncome, energyExpense, energyShare, energySent = spGetTeamResources(myTeamID, "energy")
+
+		-- How much metal and energy are builders pulling? (This number will drop when they become resource-stalled, perhaps due to being low-priority.)
+		cacheDataBase['usedBPMetalExpense'] = cacheDataBase['usedBPMetalExpense'] + usedBPMetalExpense
+		cacheDataBase['usedBPEnergyExpense'] = cacheDataBase['usedBPEnergyExpense'] + usedBPEnergyExpense
+		cacheDataBase['usedBPExceptStalled'] = cacheDataBase['usedBPExceptStalled'] + nonStalledBuildingBP
+		cacheDataBase['usedBPIfNoStall'] = cacheDataBase['usedBPIfNoStall'] + buildingBP
+
+		if config.drawBPWindRangeIndicators then
+			local realWindStrength = 0
+			for unitID in pairs(trackedWinds) do
+				local metalMake, metalUse, energyMake, energyUse = spGetUnitResources(unitID)
+				if energyMake ~= nil then
+					realWindStrength = energyMake
+					break
+				end
+			end
+			cacheDataBase['realWindStrength'] = realWindStrength
```

Comment:
Id love if this data tracking could be extracted into an API widget and made available to other widgets! Not just this value, but any others. (Even the values that arent kept 100% up-to-date every frame they could be useful.)